File: 948. Bag of Tokens.java
class Solution {
    public int bagOfTokensScore(int[] tokens, int power) {
        Arrays.sort(tokens);
        int answer = 0;
        int score = 0;
        int left = 0;
        int right = tokens.length-1;
        while (left <= right) {
            if (tokens[left] <= power) {
                power -= tokens[left++];
                score++;
                answer = Math.max(score, answer);
            } else {
                if (score == 0) return 0;
                power += tokens[right--];
                score--;
            }
        }
        return answer;
    }
}

File: go/948. Bag of Tokens.go
func bagOfTokensScore(tokens []int, power int) int {
    slices.Sort(tokens)
    left, right, answer, score := 0, len(tokens)-1, 0, 0
    for left <= right {
        if power >= tokens[left] {
            power -= tokens[left]
            left++
            score++
            answer = max(answer, score)
        } else {
            if score == 0 {
                return 0
            }
            power += tokens[right]
            right--
            score--
        }
    }
    return answer
}


